src/QueryResults.ts

```diff
@@ -0,0 +1,62 @@
+import Promise, { Thenable, Executor } from 'dojo-core/Promise';
+import Task from 'dojo-core/async/Task';
+import {duplicate, mixin} from 'dojo-core/lang';
+import * as dstore from './interfaces';
+
+export interface QueryOptions {
+	totalLength?: number| Promise<number>;
+	response: any;
+	length?: number;
+	forEach?: <T>(callback:(value: T, index: number, source: T[]) => any, instance: any) => Promise<void>;
+}
+
+function forEach<T>(callback: (value: T, index?: number, array?: T[]) => void, thisObject?: any): Promise<void> {
+	const execute: Executor<T[]> = function (resolve, reject) {
+		if (this.then) {
+			this.then(resolve, reject);
+		}
+		else {
+			resolve(this);
+		}
+	};
+	return  new Promise<Array<T>>(execute.bind(this)).then(function (data) {
+		for (let i = 0, l = data.length; i < l; i++) {
+			callback.call(thisObject, data[i], i, data);
+		}
+	});
+}
+
+export default function QueryResults<T>(data: dstore.FetchPromise<T>| Task<T> | T[], options: QueryOptions):
+	dstore.FetchPromise<T> | T[] {
+	const hasTotalLength = options && 'totalLength' in options;
+	let resultsData: dstore.FetchPromise<T>;
+	/* TODO - Remove (<any> data.then and cast, only needed because we're interacting with Dojo 1 Promises */
+	if (data instanceof Promise || (<any> data).then) {
+		resultsData = new Promise(function (resolve, reject) {
+			return (<dstore.FetchPromise<T>> data).then(resolve, reject);
+		});
+		// a promise for the eventual realization of the totalLength, in
+		// case it comes from the resolved data
+		const totalLengthPromise = resultsData.then(function (data: T[]) {
+			// calculate total length, now that we have access to the resolved data
+			let totalLength = hasTotalLength ? options.totalLength :
+				(<any> data).totalLength || data.length;
+			// make it available on the resolved data
+			(<any> data).totalLength = totalLength;
```

Comment:
I suspect it may be possible to avoid these any casts with another interface, but it might become irrelevant if we manage to further simplify this module later.